File: app/src/main/java/com/nicholasdoglio/eyebleach/di/DatabaseModule.java
/*
    Aww Gallery
    Copyright (C) 2017  Nicholas Doglio

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nicholasdoglio.eyebleach.di;

import android.app.Application;
import android.arch.persistence.room.Room;

import com.nicholasdoglio.eyebleach.data.source.local.RedditPostDatabase;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * @author Nicholas Doglio
 */
@Module
public class DatabaseModule {
    @Provides
    @Singleton
    RedditPostDatabase provideRoom(Application application) {
        return Room.databaseBuilder(application, RedditPostDatabase.class, "reddit_posts_db")
                .build();
    }
}

File: app/src/main/java/com/nicholasdoglio/eyebleach/di/DatabaseModule.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Nicholas Doglio
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.nicholasdoglio.eyebleach.di

import android.app.Application
import com.nicholasdoglio.eyebleach.Database
import com.nicholasdoglio.eyebleach.db.RedditPostQueries
import com.squareup.sqldelight.android.AndroidSqliteDriver
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
object DatabaseModule {

    private const val REDDIT_POST = "reddit_posts_db"

    @Provides
    @Singleton
    fun driver(app: Application): AndroidSqliteDriver = AndroidSqliteDriver(Database.Schema, app, REDDIT_POST)

    @Provides
    @Singleton
    fun database(driver: AndroidSqliteDriver): Database = Database(driver)

    @Provides
    fun postQueries(database: Database): RedditPostQueries = database.redditPostQueries
}
